### PR TITLE
[DDP] Use NCCL allocated memory for gradient bucket

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146589

So that NVLink SHARP comes with zero-copy on H100+ platforms, for DDP applications.
Less SM usage, less memory contention between NCCL kernel and compute kernels.

Added env `DDP_DISABLE_COMM_MEM` as a back-out option:
```
An environment variable to disable comm-optimized memory pool.
Default is 0, which means comm-optimized memory pool is enabled.
Users can set it to 1 in case of seeing regression or OOM (because this
comm MemPool may not share space with regular compute MemPool).
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D69297766](https://our.internmc.facebook.com/intern/diff/D69297766)

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+  at::cuda::OptionalCUDAGuard gpuGuard(device);
+
+  // Create memory pool
+  if (!memPool_) {
+    // Needs a CUDAAllocator
+    auto allocator =
+        reinterpret_cast<c10::cuda::CUDACachingAllocator::CUDAAllocator*>(
+            getMemAllocator().get());
+    // Pool is created
+    memPool_ = new c10::cuda::MemPool(allocator);
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5447]: Should we explicitly use `new` and delete or would handling it with a unique_ptr be better?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 2]
```diff
@@ -1294,6 +1301,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // Internal cached value: use NCCL non-blocking API mode or not.
   // Use `useNonblocking()` method instead of accessing this variable directly.
   std::optional<bool> useNonblocking_{std::nullopt};
+
+  // Communication-optimized memory pool associated with this PG
+  c10::cuda::MemPool* memPool_ = nullptr;
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:1306]: Why not have it be a std::unique_ptr?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -1543,6 +1543,8 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
     LOG(INFO) << logPrefix()
               << "ProcessGroupNCCL onCompletionHookThread thread joined.";
   }
+
+  delete memPool_;
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1547]: Use a uniquePtr remove error prone deletes

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 4]
```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
   return ncclMemAllocator;
 }
 
+at::Tensor ProcessGroupNCCL::allocateTensor(
+    long size,
+    at::TensorOptions options) {
+  // Some checks
+  TORCH_CHECK(options.has_device(), "Tensor options must include device");
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5432]:
```suggestion
  TORCH_CHECK_VALUE(options.has_device(), "Tensor options must include device");
```

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 5]
```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+  auto device = options.device();
+  TORCH_CHECK(
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5434]:
```suggestion
  TORCH_CHECK_VALUE(
```

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5447]: Yeah, that's a good idea. I used `unique_ptr` originally but realized that almost all APIs around `MemPool` takes raw pointers -- so I had to use a lot of `.get()`. Any recommendation for this situation if we use `unique_ptr`?

Comment (Skylion007) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5447]: @kwen2501 Annoying for the get() calls, but still better than manual memory management for a whole host of reasns.

Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5447]: Thanks!

Comment (syed-ahmed) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp]: `unique_ptr` looks good to me because ProcessGroupNCCL has ownership of the pool.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 6]
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+  // Create memory pool
+  if (!memPool_) {
+    // Needs a CUDAAllocator
+    auto allocator =
+        reinterpret_cast<c10::cuda::CUDACachingAllocator::CUDAAllocator*>(
+            getMemAllocator().get());
+    // Pool is created
+    memPool_ = std::make_unique<c10::cuda::MemPool>(allocator);
+    // Also need to ncclCommRegister it so that NCCL recognizes it
+    registerMemPool(memPool_.get());
```
Comment (syed-ahmed) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5447]: This call should happen *after* the tensor is allocated. Inside `registerMemPool`, we call snapshot, which returns all the blocks held by the pool. The pointers of those blocks are then registered with NCCL via ncclCommRegister. So right now, this is registering nothing.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 7]
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+    LOG(INFO) << logPrefix() << "Created memory pool";
+  }
+
+  // Allocate tensor under this MemPool's context
```
Comment (syed-ahmed) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5451]: Add a `deregisterMemPool` call here, otherwise we'll hit this `TORCH_CHECK`: https://github.com/pytorch/pytorch/blob/33191bb664fef338d94586a05bf1f14d17a00340/torch/csrc/distributed/c10d/NCCLUtils.hpp#L477-L482

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 8]
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+  auto ctx = c10::cuda::MemPoolContext(memPool_.get());
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(
+      memPool_->device(), memPool_->id(), [](cudaStream_t) { return true; });
```
Comment (syed-ahmed) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5454]: Add a `deregisterMemPool` call here, otherwise we'll hit this `TORCH_CHECK`: https://github.com/pytorch/pytorch/blob/33191bb664fef338d94586a05bf1f14d17a00340/torch/csrc/distributed/c10d/NCCLUtils.hpp#L477-L482

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 9]
```diff
@@ -5423,6 +5423,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+  at::Tensor tensor = at::empty({size}, options);
```
Comment (syed-ahmed) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5455]: Add the `registerMemPool(memPool_.get());` here. After that, all pointers in the pool will get registered.

**Review by syed-ahmed — CHANGES_REQUESTED** [review]
Thanks for the PR! Left a couple comments on registration usage. basically, 
```
deregisterMemPool(...)
at::Tensor tensor = ...
registerMemPool(...)
```

Otherwise, LGTM!

**Review by syed-ahmed — APPROVED** [review]

**Comment by kwen2501** [comment]
Thanks @syed-ahmed for the comments! I totally missed the segment level in the MemPool! I wonder if there is a way to get incremental segments? (so that we just register the new ones)
Maybe diff'ing two snapshots?

**Comment by kwen2501** [comment]
Perhaps we can maintain a segment-address-to-bool dictionary to record the registration status of segments?

File: test/distributed/test_c10d_nccl.py [hunk 10]
```diff
@@ -1937,7 +1937,8 @@ def first_bucket_size(ddp_bucket_mb):
                         opt = torch.optim.SGD(m.parameters(), lr=0.1)
                         opt_ddp = torch.optim.SGD(m_ddp.parameters(), lr=0.1)
                         has_half = any(p.dtype is torch.half for p in m.parameters())
-                        tol = 1.0e-3 if has_half else 1.0e-5
+                        atol = 1.0e-3 if has_half else 1.0e-4
+                        rtol = 1.0e-3
```
Comment (yifuwang) [inline test/distributed/test_c10d_nccl.py:1941]: Hmm, reduction over registered buffer should lead to better precision if using NVLS, and at least should not be worse than the non-registered baseline if using ring. Do you know why this needs to be relaxed?

Comment (kwen2501) [reply test/distributed/test_c10d_nccl.py:1941]: In general yes, but I am not 100% sure about a two-rank case (if that's the world size of this test)

**Comment by syed-ahmed** [comment]
> Perhaps we can maintain a segment-address-to-bool dictionary to record the registration status of segments? (Edit: I see that registeredSegmentHandles_ already does so)

Yeah, may be we can make `registerSegment` a no-op when a segment is already registered. However, that function was originally implemented for the `TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK` path: https://github.com/pytorch/pytorch/blob/3f5ed0568886ca4c144213a27918019a2e3ebab0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L311-L328. So not sure if it would break any internal workload.

File: test/distributed/test_c10d_nccl.py (hunk 10 above)
Comment (yifuwang) [reply test/distributed/test_c10d_nccl.py:1941]: Hmm, IIUC, with rank=2, floating point addition shouldn't suffer from associativity problem at all. Which means we should expect identical results?

**Comment by kwen2501** [comment]
Yeah, I am adding a flag `errorOnRereg` to `registerSegment` API with a default value of true. In `registerMemPool`, we can set the flag to false.

**Comment by syed-ahmed** [comment]
We should also add a `deregisterMemPool` in ProcessGroupNCCL::shutdown(), and I think that will make this diff complete :).

**Comment by kwen2501** [comment]
I left `deregisterMemPool` out because I wasn't sure when it would be safe to call it --
- If we call it before `ncclCommFinalize`, at that point the collectives are still executing, would the dereg create any side effect? (not very likely but I cannot rule it out)
- If we call it after `ncclCommFinalize`, at that point could some resources of the communicator be freed already and cause the dereg call to fail? 
I'd need more confirmation before adding in that call.

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 11]
```diff
@@ -1157,14 +1157,37 @@ void Reducer::initialize_buckets(
...
+        // `GLOO`. In this case, we just fall back to the regular way of
+        // creating tensor
+        LOG(INFO)
+            << "Reducer: default comm backend not found, skipping bucket memory optimization";
+      }
+      if (backend != nullptr && backend->supportsTensorAlloc()) {
+        // Comm-optimized memory pool is available, use it to allocate tensor
+        LOG(INFO)
+            << "Reducer: found comm-optimized memory allocator, using it to create bucket";
+        bucket.gradients = backend->allocateTensor(bucketSize, options);
```
Comment (fegin) [inline torch/csrc/distributed/c10d/reducer.cpp:1184]: noob question, is the available NCCL available memory size the same? If it is smaller, can it lead to easier to OOM?

**Comment by syed-ahmed** [comment]
Looking at NCCL source, it looks like `ncclCommDeregister` is just cleaning up some cache. Also `ncclCommDestroy` calls the cleanup function as well. I think it's safe to put it after `ncclCommFinalize` as `ncclCommDestroy` happens after `ncclCommFinalize`.
```
ncclCommDeregister -> commDeregister -> regCleanup
ncclCommDestroy -> commReclaim -> commCleanup -> commFree -> ncclRegCleanup -> regCleanup

```

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 11 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/reducer.cpp:1184]: The NCCL memory pool does not have a limited size per my understanding. It is extendable as like regular CUDA caching allocator.

**Comment by kwen2501** [comment]
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69297766).

**Comment by kwen2501** [comment]
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69297766).

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 12]
```diff
@@ -1157,14 +1157,44 @@ void Reducer::initialize_buckets(
...
+      try {
+        backend = process_group_->getDefaultBackend();
+      } catch (...) {
+        // Sometimes the backend type can be `UNDEFINED` rather than `NCCL` or
+        // `GLOO`. In this case, we just fall back to the regular way of
+        // creating tensor
+        LOG(INFO)
+            << "Reducer: default comm backend not found, skipping bucket memory optimization";
+      }
+      if (ddpDisableCommMem == 0 && backend != nullptr &&
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/reducer.cpp:1186]: Do we like `ddpDisableCommMem == 0` or `!ddpDisableCommMem` (because the var is a bool) or it doesn't matter?
just a clarifying question - if in pytorch do we have a style guide for this?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 13]
```diff
@@ -5423,6 +5432,46 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
   return ncclMemAllocator;
 }
 
+at::Tensor ProcessGroupNCCL::allocateTensor(
+    long size,
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5436]: is there a max/recommended size limit?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 14]
```diff
@@ -1157,14 +1157,44 @@ void Reducer::initialize_buckets(
...
+      // Allocate the bucket's flattened `gradients` tensor.
+      auto bucketSize = static_cast<long>(offset);
+      // Check if we can use comm-optimized memory pool to allocate tensor
+      c10::intrusive_ptr<Backend> backend = nullptr;
+      // An environment variable to disable comm-optimized memory pool.
+      // Default is 0, which means comm-optimized memory pool is enabled.
+      // Users can set it to 1 in case of seeing regression or OOM (because this
+      // comm MemPool may not share space with regular compute MemPool).
+      bool ddpDisableCommMem =
+          (getCvarString({"DDP_DISABLE_COMM_MEM"}, "0") == "1");
```
Comment (c-p-i-o) [inline torch/csrc/distributed/c10d/reducer.cpp:1176]: do you think it's worth logging the value of this env variable in `ProcessGroupNCCL` so we can search for it in logs?

**Review by c-p-i-o — APPROVED** [review]

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 14 above)
Comment (fduwjj) [reply torch/csrc/distributed/c10d/reducer.cpp:1176]: Or ddp log since this is part of DDP.

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 15]
```diff
@@ -1157,14 +1157,44 @@ void Reducer::initialize_buckets(
...
       if (mixed_precision_param_dtype_.has_value()) {
         options = options.dtype(mixed_precision_param_dtype_);
       }
-      bucket.gradients = at::empty({static_cast<long>(offset)}, options);
+
+      // Allocate the bucket's flattened `gradients` tensor.
+      auto bucketSize = static_cast<long>(offset);
+      // Check if we can use comm-optimized memory pool to allocate tensor
+      c10::intrusive_ptr<Backend> backend = nullptr;
+      // An environment variable to disable comm-optimized memory pool.
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/reducer.cpp:1171]: It would be great that we can explain in the comment why or what benefit of doing this. What you wrote in the summary of this PR is a good fit for it.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 16]
```diff
@@ -5423,6 +5432,46 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
...
+  // Allocate tensor under this MemPool's context
+  auto ctx = c10::cuda::MemPoolContext(memPool_.get());
+  c10::cuda::CUDACachingAllocator::beginAllocateToPool(
+      memPool_->device(), memPool_->id(), [](cudaStream_t) { return true; });
+  at::Tensor tensor = at::empty({size}, options);
+  // Also need to ncclCommRegister the pool in case new segments are created;
+  // reregistration of old segments will be ignored
+  registerMemPool(memPool_.get());
+  c10::cuda::CUDACachingAllocator::endAllocateToPool(
+      memPool_->device(), memPool_->id());
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:5467]: Will this cause fragmentation?

**Review by fduwjj — APPROVED** [review]
unblock and in general looks good. I have one more suggestion for Meta internal use case which will follow up with you offline.

**Comment by Skylion007** [comment]
Do we need to do anything, should we do anything for other NCCL comms like all_gather, all_scatter, etc?
